key-rotator/key/key.go

```diff
@@ -98,52 +145,51 @@ func (cfg RotationConfig) Validate() error {
 //    `create_min_age`, create a new key version.
 //  * While there are more than `delete_min_key_count` keys, and the oldest key
 //    version is older than `delete_min_age`, delete the oldest key version.
-//  * Mark a single key version as primary (unmarking any other key versions
-//    that may be marked primary):
-//    * If there is a key version older than `primary_min_age`, mark the
-//      youngest such key version as primary.
-//    * Otherwise, mark the oldest key version as primary.
+//  * Determine the current primary version:
+//    * If there is a key version not younger than `primary_min_age`, select
+//      the youngest such key version as primary.
+//    * Otherwise, select the oldest key version as primary.
+//
+// The returned key is guaranteed to include at least one version.
 func (k Key) Rotate(now time.Time, cfg RotationConfig) (Key, error) {
 	// Validate parameters.
 	if err := cfg.Validate(); err != nil {
 		return Key{}, fmt.Errorf("invalid rotation config: %w", err)
 	}
 
-	// Copy the existing list of key versions, sorting by creation time. Also,
-	// validate that we aren't trying to rotate a key containing a version from
-	// the "future" to simplify later logic, and go ahead and unmark primary on
-	// all key versions so that we can easily mark a single version primary
-	// later.
-	age := func(v Version) time.Duration { return now.Sub(v.CreationTime) }
-	kvs := make([]Version, 0, 1+len(k.v))
+	// Copy the existing list of key versions, sorting by creation time
+	// ascending (oldest to youngest). Also, validate that we aren't trying to
+	// rotate a key containing a version from the "future" to simplify later
+	// logic.
+	nowTS := now.Unix()
+	age := func(v Version) time.Duration { return time.Second * time.Duration(nowTS-v.CreationTimestamp) }
+	vs := make([]Version, 0, 1+len(k.v))
 	for _, v := range k.v {
 		if age(v) < 0 {
-			return Key{}, fmt.Errorf("found key version with creation time %v, after now (%v)", v.CreationTime.Format(time.RFC3339), now.Format(time.RFC3339))
+			return Key{}, fmt.Errorf("found key version with creation timestamp %d, after now (%d)", v.CreationTimestamp, nowTS)
 		}
-		v.Primary = false
-		kvs = append(kvs, v)
+		vs = append(vs, v)
 	}
-	sort.Slice(kvs, func(i, j int) bool { return kvs[i].CreationTime.Before(kvs[j].CreationTime) })
+	sort.Slice(vs, func(i, j int) bool { return vs[i].CreationTimestamp < vs[j].CreationTimestamp })
 
 	// Policy: if no key versions exist, or if the youngest key version is
 	// older than `create_min_age`, create a new key version.
-	if len(kvs) == 0 || age(kvs[len(kvs)-1]) > cfg.CreateMinAge {
+	if len(vs) == 0 || age(vs[len(vs)-1]) > cfg.CreateMinAge {
```

Comment:
Nit:

```
age(vs[len(vs)-1])
```

To me this is a bit hard to follow what's happening there.

Reply:
I added some clarifying comments, and pulled out the inner indexing expression onto another line.